ducktape/cluster/node_container.py

```diff
@@ -112,16 +117,43 @@ def remove_spec(self, cluster_spec):
         :throws InsufficientResourcesError:     If there are not enough nodes in the NodeContainer.
                                                 Nothing will be removed unless enough are available.
         """
-        msg = self.attempt_remove_spec(cluster_spec)
-        if len(msg) > 0:
-            raise InsufficientResourcesError("Not enough nodes available to allocate. " + msg)
-        removed = []
+        err = ""
+        good = []
+        bad = []
         for os, node_specs in iteritems(cluster_spec.nodes.os_to_nodes):
             num_nodes = len(node_specs)
+            good_per_os = []
+            bad_per_os = []
             avail_nodes = self.os_to_nodes.get(os, [])
-            for i in range(0, num_nodes):
-                removed.append(avail_nodes.pop(0))
-        return removed
+            if len(avail_nodes) < num_nodes:
+                err = err + "%s nodes requested: %d. Total %s nodes available: %d" % \
+                  (os, num_nodes, os, len(avail_nodes))
+            else:
+                for i in range(0, num_nodes):
+                    node = avail_nodes.pop(0)
+                    if isinstance(node, RemoteAccount):
+                        try:
+                            if node.available():
+                                good_per_os.append(node)
+                            else:
+                                bad_per_os.append(node)
+                        finally:
+                            node.close()
+                    else:
+                        good_per_os.append(node)
+
+                bad.extend(bad_per_os)
+                # if we don't have enough good linux/windows nodes to allocate,
+                # report it and don't actually allocate anything
+                if len(good_per_os) < num_nodes:
+                    err = err + "%s nodes requested: %d. Healthy %s nodes available: %d" % \
+                      (os, num_nodes, os, len(good_per_os))
+                    for node in good_per_os:
+                        self.add_node(node)
+                else:
+                    good.extend(good_per_os)
+
+        return good, bad, err
```

Comment:
id return a named tupple here for cleanliness, 

```suggestion 
    GoodAndBadNodes = namedtupple("GoodAndBadNodes", ["good", "bad"])
    ...
        return GoodAndBadNodes(good, bad)
```
that way as a user, you can do
```
good, bad = nodes.remove_spec(spec)
bad = nodes.remove_spec(spec).bad
```

Reply:
See my other comment, I'll try to wrap this in some kind of a Result object, which would have:
- good
- bad
- success: bool
- message

Cause we need to still return bad nodes if allocation failed, it can't just be an exception.

Reply:
You can return the bad nodes in the exception. 
```
class BadNodeError(InsufficientResourcesError):
    def __init__(self, *args, **kwargs):
      self.node_data = good_and_bad_nodes
   
```
that way you can drop the success variable and be a bit more pythonic

Reply:
So I took a quick stab at it, and it turns out we have to return bad nodes in both success and failure cases, leading to smth like this, also not super elegant - not sure which one I dislike more lol:
```python
# in json.py
        try:
            r: RemoveSpecResult = self._available_accounts.remove_spec(cluster_spec)
        except BadNodesError as be:
            self._bad_accounts.add_nodes(be)
            raise be

        if r.bad_nodes:
            self._bad_accounts.add_nodes(r.bad_nodes)
```
we could probably do without even saving bad nodes anywhere, but I was actually thinking about logging them out after the test run is over.

Reply:
So after contemplating it some more I did change it to use exceptions, it does look a bit cleaner overall. PTAL

Reply:
cool!